tests/integration/tests/charm_tests/openstack_exporter.py

```diff
@@ -256,18 +257,45 @@ def test_openstack_exporter_snap_down(self):
         )
         self.assertEqual(self.leader_unit.workload_status_message, "")
 
+    def test_upstream_exporter_as_resource(self):
+        """Test using the upstream golang as resource will block the unit."""
+        # juju cannot access resources at /tmp, so we create a tmp folder at the current directory
+        with tempfile.TemporaryDirectory(dir="./") as tmpdir:
+            tmp_path = Path(tmpdir)
+            download_cmd = (
+                f"wget -q -P {tmpdir} https://github.com/canonical/openstack-exporter-operator/"
+                "releases/download/rev2/golang-openstack-exporter_amd64.snap"
+            )
+            subprocess.run(download_cmd.split(), check=False)
+
+            resource_file = tmp_path / "golang-openstack-exporter_amd64.snap"
+            model.attach_resource(APP_NAME, "openstack-exporter", str(resource_file))
+            model.block_until_unit_wl_status(
+                self.leader_unit_entity_id, "blocked", timeout=STATUS_TIMEOUT
+            )
+            snaps_installed = [snap["name"] for snap in get_snaps_installed()]
+            self.assertIn(UPSTREAM_SNAP, snaps_installed)
+            expected_msg = (
+                f"{UPSTREAM_SNAP} should not be used anymore. " "Please remove the snap resource."
```

Comment:
By looking at this block message and the necessary step below, I believe it's worth an upgrade document. And the wording "please remove" may be misleading if Juju doesn't have a way to **remove** the resource. It would be good to suggest "by uploading an empty resource" or something like that.

Reply:
I've changed the wording to add an empty file and pointing also to the [known-issues](https://charmhub.io/openstack-exporter#known-issues) charmhub page.

We'll update the [tutorial](https://charmhub.io/openstack-exporter/docs/tutorial#openstack-exporter-snap) to not use the resource when this changes land into stable.

Reply:
docs/tutorial updates are now live

Reply:
I added a follow up comment https://github.com/canonical/openstack-exporter-operator/pull/93#discussion_r1737620856